### PR TITLE
Implement RFC3695 Allow boolean literals as cfg predicates

### PR DESCRIPTION
**Pull request by Urgau**
This PR implements https://github.com/rust-lang/rfcs/pull/3695: allow boolean literals as cfg predicates, i.e. `cfg(true)` and `cfg(false)`.

r? @nnethercote *(or anyone with parser knowledge)*
cc @clubby789

### CODE REVIEW

File: compiler/rustc_feature/src/unstable.rs [hunk 1]
```diff
@@ -371,6 +371,8 @@ declare_features! (
     (unstable, async_for_loop, "1.77.0", Some(118898)),
     /// Allows using C-variadics.
     (unstable, c_variadic, "1.34.0", Some(44930)),
+    /// Allows the use of `#[cfg(<true/false>)]`.
+    (unstable, cfg_boolean_literal, "CURRENT_RUSTC_VERSION", Some(999999)),
```
Comment (Urgau) [inline compiler/rustc_feature/src/unstable.rs:375]: note to self: remember to update the tracking issue number here (and the the unstable-book) once the RFC is merged and the tracking issue is created.

File: compiler/rustc_builtin_macros/src/cfg.rs [hunk 2]
```diff
@@ -36,14 +35,18 @@ pub(crate) fn expand_cfg(
...
+    tts: TokenStream,
+) -> PResult<'a, ast::NestedMetaItem> {
     let mut p = cx.new_parser_from_tts(tts);
 
     if p.token == token::Eof {
         return Err(cx.dcx().create_err(errors::RequiresCfgPattern { span }));
     }
 
-    let cfg = p.parse_meta_item(AllowLeadingUnsafe::Yes)?;
+    let cfg = p.parse_meta_item_inner()?;
```
[the remark this answers is not in the document]
Comment (Urgau) [reply compiler/rustc_builtin_macros/src/cfg.rs:49]: Well to be honest, I think the `AllowLeadingUnsafe::Yes` for `cfg!` is a mistake, as it allows `cfg!(unsafe(foo))`, which makes no sense.

[the remark this answers is not in the document]
Comment (jieyouxu) [reply compiler/rustc_builtin_macros/src/cfg.rs:49]: I mean fair enough, though we should demonstrate that this is intentional change with a test case if one does not already exist.

[the remark this answers is not in the document]
Comment (Urgau) [reply compiler/rustc_builtin_macros/src/cfg.rs:49]: I will take a look at the other `AllowLeadingUnsafe` usages and send a PR fix (with test case) and beta-backport (it seems).

File: tests/ui/macros/cfg.stderr [hunk 3]
```diff
@@ -6,11 +6,11 @@ LL |     cfg!();
    |
    = note: this error originates in the macro `cfg` (in Nightly builds, run with -Z macro-backtrace for more info)
 
-error: expected identifier, found `123`
+error[E0565]: unsupported literal
```
Comment (nnethercote) [inline tests/ui/macros/cfg.stderr:9]: "unsupported literal" is a very basic error message, and doesn't give any idea what kind of literal is allowed. There is `attr_unsupported_literal_cfg_string`, which becomes "literal in `cfg` predicate value must be a string". Could something like that be used to indicate only a boolean literal is allowed?

File: compiler/rustc_trait_selection/src/error_reporting/traits/on_unimplemented.rs [hunk 4]
```diff
@@ -414,7 +414,7 @@ impl<'tcx> OnUnimplementedDirective {
             let cond = item_iter
                 .next()
                 .ok_or_else(|| tcx.dcx().emit_err(EmptyOnClauseInOnUnimplemented { span }))?
-                .meta_item()
+                .meta_item_or_bool()
```
Comment (nnethercote) [inline compiler/rustc_trait_selection/src/error_reporting/traits/on_unimplemented.rs:417]: This method is used in the first commit but only defined in the second commit, which is a little strange when reading the commits one at a time.

File: tests/ui/cfg/true-false.rs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+#[cfg_attr(true, cfg(false))]
+fn foo() {}
+
+#[link(name = "foo", cfg(false))]
+extern "C" {}
+
+fn main() {
+    assert!(foo());
```
Comment (nnethercote) [inline tests/ui/cfg/true-false.rs:23]: Should this be `!foo()`?

File: tests/ui/cfg/true-false.rs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    assert!(!cfg!(false));
+    assert!(cfg!(not(false)));
+    assert!(cfg!(all(true)));
+    assert!(cfg!(any(true)));
+    assert!(cfg!(any(false, true)));
+    assert_eq!(cfg!(true), cfg!(not(false)));
```
Comment (nnethercote) [inline tests/ui/cfg/true-false.rs:29]: Just two separate `assert!`s would be simpler and more consistent with what's above.

File: src/doc/unstable-book/src/language-features/cfg-boolean-literals.md [hunk 7]
```diff
@@ -0,0 +1,22 @@
+# `cfg_boolean_literals`
+
+The tracking issue for this feature is: [#00000]
+
+[#00000]: https://github.com/rust-lang/rust/issues/00000
```
Comment (nnethercote) [inline src/doc/unstable-book/src/language-features/cfg-boolean-literals.md:5]: Is updating this number part of the "note to self" above?

File: compiler/rustc_parse/src/parser/attr.rs [hunk 8]
```diff
@@ -452,7 +454,7 @@ impl<'a> Parser<'a> {
     /// ```ebnf
     /// MetaItemInner = UNSUFFIXED_LIT | MetaItem ;
     /// ```
-    fn parse_meta_item_inner(&mut self) -> PResult<'a, ast::NestedMetaItem> {
+    pub fn parse_meta_item_inner(&mut self) -> PResult<'a, ast::NestedMetaItem> {
```
Comment (nnethercote) [inline compiler/rustc_parse/src/parser/attr.rs:457]: An aside: `NestedMetaItem` would be better named as `MetaItemInner`, to match the grammar.

**Comment by nnethercote** [comment]
Basically looks good, a few comments above.

File: tests/ui/cfg/true-false.rs (hunk 5 above)
Comment (Urgau) [reply tests/ui/cfg/true-false.rs:23]: No, `foo()` does return `true` from the first definition:
```
#[cfg(true)]
fn foo() -> bool {
    cfg!(true) // this return true
}
```

File: compiler/rustc_parse/src/parser/attr.rs (hunk 8 above)
Comment (Urgau) [reply compiler/rustc_parse/src/parser/attr.rs:457]: Agree, at first I was bit confused about `MetaItemInner` vs `NestedMetaItem`, until realizing that they are the same.

File: src/doc/unstable-book/src/language-features/cfg-boolean-literals.md (hunk 7 above)
Comment (Urgau) [reply src/doc/unstable-book/src/language-features/cfg-boolean-literals.md:5]: Yes, "and in the unstable-book".

File: tests/ui/macros/cfg.stderr (hunk 3 above)
Comment (Urgau) [reply tests/ui/macros/cfg.stderr]: Yes, see the last commit [`f9fc1fb` (#131034)](https://github.com/rust-lang/rust/pull/131034/commits/f9fc1fb5c8af2077328dfbe849d7608189818877).

File: compiler/rustc_parse/src/parser/attr.rs (hunk 8 above)
Comment (Urgau) [reply compiler/rustc_parse/src/parser/attr.rs:457]: Opened https://github.com/rust-lang/rust/issues/131087 to track this renaming.

**Comment by nnethercote** [comment]
Thanks!

@bors r+

**Comment by Urgau** [comment]
Let's wait until the RFC is actually merged, so that I can put the right tracking issue number. I will r=you when it's done.

@bors r-

File: compiler/rustc_feature/src/unstable.rs (hunk 1 above)
Comment (clubby789) [reply compiler/rustc_feature/src/unstable.rs]: I created a tracking issue (#131204) since it seems like this will be merged soon

**Comment by Urgau** [comment]
Added the tracking issue. CI passes. Let's go.

@bors r=@nnethercote

**Comment by Urgau** [comment]
>    = note: LINK : fatal error LNK1104: cannot open file 'C:\a\rust\rust\build\bootstrap\debug\deps\rustc.exe'

@bors retry fatal error LNK1104 #127883

**Pull request merged**